lxd/seccomp/seccomp.go

```diff
@@ -249,6 +259,165 @@ static void prepare_seccomp_iovec(struct iovec *iov,
 	iov[3].iov_len = SECCOMP_COOKIE_SIZE;
 }
 
+static inline int pidfd_getfd(int pidfd, int fd, int flags)
+{
+	return syscall(__NR_pidfd_getfd, pidfd, fd, flags);
+}
+
+#define ptr_to_u64(p) ((__aligned_u64)((uintptr_t)(p)))
+
+static inline int bpf(int cmd, union bpf_attr *attr, size_t size)
+{
+	return syscall(__NR_bpf, cmd, attr, size);
+}
+
+static int handle_bpf_syscall(int notify_fd, int mem_fd, struct seccomp_notify_proxy_msg *msg,
+			      struct seccomp_notif *req, struct seccomp_notif_resp *resp,
+			      int *bpf_cmd, int *bpf_prog_type, int *bpf_attach_type)
+{
+	__do_close int pidfd = -EBADF, bpf_target_fd = -EBADF, bpf_attach_fd = -EBADF,
+		       bpf_prog_fd = -EBADF;
+	__do_free char *log_buf = NULL;
+	__do_free struct bpf_insn *insn = NULL;
+	char license[128];
+	size_t insn_size;
+	union bpf_attr attr = {}, new_attr;
+	unsigned int attr_len = sizeof(attr);
+	struct seccomp_notif_addfd addfd = {};
+	int ret;
+	int cmd;
+
+	*bpf_cmd = -EINVAL;
+	*bpf_prog_type = -EINVAL;
+	*bpf_attach_type = -EINVAL;
+
+	if (attr_len < req->data.args[2])
+		return -EFBIG;
+	attr_len = req->data.args[2];
+
+	*bpf_cmd = req->data.args[0];
+	switch (req->data.args[0]) {
+	case BPF_PROG_LOAD:
+		cmd = BPF_PROG_LOAD;
+		break;
+	case BPF_PROG_ATTACH:
+		cmd = BPF_PROG_ATTACH;
+		break;
+	case BPF_PROG_DETACH:
+		cmd = BPF_PROG_DETACH;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	ret = pread(mem_fd, &attr, attr_len, req->data.args[1]);
+	if (ret < 0)
+		return -errno;
+
+	*bpf_prog_type = attr.prog_type;
+
+	pidfd = pidfd_open(req->pid, 0);
+	if (pidfd < 0)
+		return -errno;
+
+	if (ioctl(notify_fd, SECCOMP_IOCTL_NOTIF_ID_VALID, &req->id))
+		return -errno;
+
+	switch (cmd) {
+	case BPF_PROG_LOAD:
+		if (attr.prog_type != BPF_PROG_TYPE_CGROUP_DEVICE)
+			return -EINVAL;
+
+		insn_size = sizeof(struct bpf_insn) * attr.insn_cnt;
+
+		insn = malloc(insn_size);
+		if (!insn)
+			return -ENOMEM;
+
+		ret = pread(mem_fd, insn, insn_size, attr.insns);
+		if (ret < 0)
+			return -errno;
+		if (ret != insn_size)
+			return -EIO;
+
+		memcpy(&new_attr, &attr, sizeof(attr));
+
+		if (attr.log_size > 0 && attr.log_size <= (UINT_MAX / 2)) {
+			log_buf = malloc(attr.log_size);
```

Comment:
Isn't it risky if the container can trigger the runtime to allocate a very big amount of memory?

Reply:
I think it's fine because it is a complete opt-in feature that the administrator needs to enable on a specific container. If the workload is completely untrusted I wouldn't enable it. Arguably ~250 mb is a bit rich as upper bound and we could just always use a static buffer of a fixed size, say 4096 bytes or a little more then this wouldn't be a problem anymore imho.  You want to send a patch? ;)